python/pyspark/rdd.py

```diff
@@ -2029,7 +2028,11 @@ def coalesce(self, numPartitions, shuffle=False):
         >>> sc.parallelize([1, 2, 3, 4, 5], 3).coalesce(1).glom().collect()
         [[1, 2, 3, 4, 5]]
         """
-        jrdd = self._jrdd.coalesce(numPartitions, shuffle)
+        if shuffle:
+            data_java_rdd = self._to_java_object_rdd().coalesce(numPartitions, shuffle)
```

Comment:
would be great to add some inline comment explaining why this is necessary. otherwise somebody can just come in 6 month from now and change this back to `jrdd = self._jrdd.coalesce(numPartitions, shuffle)`